rclc/src/rclc/action_server.c

```diff
@@ -0,0 +1,325 @@
+
+  global_goal_handle = (rclc_action_goal_handle_instance_t) {0};
+  *rclc_handle = &global_goal_handle;
+
+  (*rclc_handle)->goal_request = (Generic_SendGoal_Request*) request;
+  (*rclc_handle)->action_server = action_server;
+
+  Generic_SendGoal_Response res;
+  res.accepted = true;
+
+  rcl_ret_t rc = rcl_action_send_goal_response(action_server, request_id, &res);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_accept_request, rcl_action_send_goal_response);
+    return rc;
+  }
+
+  rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+  goal_info.goal_id = (*rclc_handle)->goal_request->goal_id;
+
+  (*rclc_handle)->rcl_handle = rcl_action_accept_new_goal(action_server, &goal_info);
+  rc = rcl_action_update_goal_state((*rclc_handle)->rcl_handle, GOAL_EVENT_EXECUTE);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_accept_request, rcl_action_update_goal_state);
+    return rc;
+  }
+
+  rcl_action_goal_status_array_t c_status_array = rcl_action_get_zero_initialized_goal_status_array();
+  rc = rcl_action_get_goal_status_array(action_server, &c_status_array);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_accept_request, rcl_action_get_goal_status_array);
+    return rc;
+  }
+
+  rc = rcl_action_publish_status(action_server, &c_status_array.msg);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_accept_request, rcl_action_publish_status);
+    return rc;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rclc_action_server_reject_goal_request(
+  rcl_action_server_t * action_server,
+  rmw_request_id_t * request_id)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    action_server, "action_server is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    request_id, "request_id is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+
+  Generic_SendGoal_Response res;
+  res.accepted = false;
+
+  rcl_ret_t rc = rcl_action_send_goal_response(action_server, request_id, &res);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(rclc_action_server_accept_request, rcl_action_send_goal_response);
+    return rc;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rclc_action_server_set_goal_result_request(
+  rclc_action_goal_handle_t rclc_handle,
+  void * ros_request,
+  rmw_request_id_t * request_header)
+{
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    ros_request, "request is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    request_header, "request_header is a null pointer", return RCL_RET_INVALID_ARGUMENT);
+
+  if(!rclc_action_server_is_valid_handle(rclc_handle)){
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  Generic_GetResult_Request * request = (Generic_GetResult_Request *) ros_request;
+
+  bool goal_id_match = 0 == memcmp(request->goal_id.uuid, rclc_handle->goal_request->goal_id.uuid, sizeof(request->goal_id.uuid));
```

Comment:
```suggestion
  bool goal_id_match = (0 == memcmp(request->goal_id.uuid, rclc_handle->goal_request->goal_id.uuid, sizeof(request->goal_id.uuid)));
```
Just to improve readability.